File: pathfinder-bukkit/src/main/java/de/cubbossa/pathfinder/command/FindLocationCommand.java
package de.cubbossa.pathfinder.command;

import de.cubbossa.pathfinder.PathFinder;
import de.cubbossa.pathfinder.PathPerms;
import de.cubbossa.pathfinder.graph.GraphEntryNotEstablishedException;
import de.cubbossa.pathfinder.graph.NoPathFoundException;
import de.cubbossa.pathfinder.messages.Messages;
import de.cubbossa.pathfinder.misc.Location;
import de.cubbossa.pathfinder.misc.PathPlayer;
import de.cubbossa.pathfinder.navigation.NavigationLocation;
import de.cubbossa.pathfinder.navigation.NavigationModule;
import de.cubbossa.pathfinder.navigation.Route;
import de.cubbossa.pathfinder.node.implementation.PlayerNode;
import de.cubbossa.pathfinder.node.implementation.Waypoint;
import de.cubbossa.pathfinder.util.BukkitUtils;
import dev.jorel.commandapi.CommandTree;
import java.util.UUID;
import java.util.concurrent.CompletionException;
import java.util.logging.Level;
import org.bukkit.entity.Player;

public class FindLocationCommand extends CommandTree {

  public FindLocationCommand() {
    super("findlocation");
    withAliases("gpslocation", "navigatelocation");
    withPermission(PathPerms.PERM_CMD_FIND_LOCATION);

    then(Arguments.location("location")
        .executesPlayer((player, args) -> {
          NavigationModule<Player> module = NavigationModule.get();
          Location target = args.getUnchecked(0);
          Waypoint waypoint = new Waypoint(UUID.randomUUID());
          waypoint.setLocation(target);

          PathPlayer<Player> p = BukkitUtils.wrap(player);
          module.setFindCommandPath(p, Route
              .from(NavigationLocation.movingExternalNode(new PlayerNode(p)))
              .to(NavigationLocation.movingExternalNode(waypoint))
          ).whenComplete((nav, throwable) -> {
            if (throwable != null) {
              if (throwable instanceof CompletionException) {
                throwable = throwable.getCause();
              }
              if (throwable instanceof NoPathFoundException) {
                p.sendMessage(Messages.CMD_FIND_BLOCKED);
              } else if (throwable instanceof GraphEntryNotEstablishedException) {
                p.sendMessage(Messages.CMD_FIND_TOO_FAR);
              } else {
                p.sendMessage(Messages.CMD_FIND_UNKNOWN);
                PathFinder.get().getLogger().log(Level.SEVERE, "Unknown error while finding path.", throwable);
              }
            }
            nav.persist().cancelWhenTargetInRange();
          });
        })
    );
  }
}

File: pathfinder-bukkit/src/main/java/de/cubbossa/pathfinder/command/FindLocationCommand.kt
package de.cubbossa.pathfinder.command

import de.cubbossa.pathfinder.PathFinder
import de.cubbossa.pathfinder.PathPerms
import de.cubbossa.pathfinder.asPathPlayer
import de.cubbossa.pathfinder.graph.GraphEntryNotEstablishedException
import de.cubbossa.pathfinder.graph.NoPathFoundException
import de.cubbossa.pathfinder.launchIO
import de.cubbossa.pathfinder.messages.Messages
import de.cubbossa.pathfinder.misc.Location
import de.cubbossa.pathfinder.navigation.NavigationLocation
import de.cubbossa.pathfinder.navigation.NavigationModule.Companion.get
import de.cubbossa.pathfinder.navigation.Route
import de.cubbossa.pathfinder.node.implementation.PlayerNode
import de.cubbossa.pathfinder.node.implementation.Waypoint
import dev.jorel.commandapi.CommandTree
import dev.jorel.commandapi.executors.PlayerCommandExecutor
import dev.jorel.commandapi.kotlindsl.locationArgument
import org.bukkit.entity.Player
import java.util.*
import java.util.logging.Level

class FindLocationCommand : CommandTree("findlocation") {
    init {
        withAliases("gpslocation", "navigatelocation")
        withPermission(PathPerms.PERM_CMD_FIND_LOCATION)

        locationArgument("location") {
            executesPlayer(
                PlayerCommandExecutor { player, args ->
                    val target: Location = args.getUnchecked(0)!!
                    val waypoint = Waypoint(UUID.randomUUID(), target)

                    val p = player.asPathPlayer()
                    launchIO {
                        val navigationModule = get<Player>()
                        try {
                            val path = navigationModule.navigate(
                                p, Route
                                    .from(NavigationLocation.movingExternalNode(PlayerNode(p)))
                                    .to(NavigationLocation.movingExternalNode(waypoint))
                            )
                            navigationModule.cancelPathWhenTargetReached(path)
                        } catch (t: NoPathFoundException) {
                            p.sendMessage(Messages.CMD_FIND_BLOCKED)
                        } catch (t: GraphEntryNotEstablishedException) {
                            p.sendMessage(Messages.CMD_FIND_TOO_FAR)
                        } catch (t: Throwable) {
                            p.sendMessage(Messages.CMD_FIND_UNKNOWN)
                            PathFinder.get().logger.log(
                                Level.SEVERE,
                                "Unknown error while finding path.",
                                t
                            )
                        }
                    }
                }
            )
        }
    }
}
